### PR TITLE
[RFC] Basic testing infrastructure

### PR DESCRIPTION
**Pull request by ngortheone**
@aaschmid 
I have created some basic testing infra for the client. It is not finished yet, but comments are welcome!

#### Why
Integration testing is important part of building high-quality client-server software 

#### What
* Dockerfile with taskwarrior server and pre-created user, to enable uniform testing setup for multiple developers
* Junit test from readme that run against dockerized taskd

TBD:
- Readme updates on how to use docker and run tests

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -8,18 +8,14 @@ plugins {
...
-version = '1.0-SNAPSHOT'
+version = '1.1-SNAPSHOT'
 
 sourceCompatibility = targetCompatibility = JavaVersion.VERSION_1_8
 
 repositories {
     mavenCentral()
 }
 
-task wrapper(type: Wrapper) {
```
Comment (ngortheone) [inline build.gradle:19]: my intellij idea barks on conflicting task. Removing this does not seem to break anything.

Comment (aaschmid) [reply build.gradle:19]: with new versions of Gradle this `task` is obsolete. So you can safely remove it. 
And if you like you can upgrade to the newest Gradle wrapper...

File: .gitignore [hunk 2]
```diff
@@ -11,3 +11,7 @@ bin/
 .project
 
 taskwarrior.properties
+
```
Comment (aaschmid) [inline .gitignore:14]: Like

File: build.gradle [hunk 3]
```diff
@@ -72,3 +76,8 @@ artifacts {
     archives javadocJar
     archives sourcesJar
 }
+
+dependencies {
+    testImplementation('org.junit.jupiter:junit-jupiter:5.5.1')
```
Comment (aaschmid) [inline build.gradle]: parentheses can be omitted -> `testImplementation 'org....'

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,54 @@
+FROM alpine:latest
```
Comment (aaschmid) [inline docker/Dockerfile:1]: LIKE

File: gradle/wrapper/gradle-wrapper.properties [hunk 5]
```diff
@@ -1,5 +1,6 @@
+#Fri Aug 30 14:22:55 EDT 2019
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.1-all.zip
```
Comment (aaschmid) [inline gradle/wrapper/gradle-wrapper.properties:2]: Good, you already upgraded :-)

**Review by aaschmid — COMMENTED** [review]
Looks good besides `docker/certs/README.md` is empty, pls fill or remove.
I also want to run it tonight.

One remaining question: Where do you want to run this automatically? travis-ci, circle-ci?

File: README.md [hunk 6]
```diff
@@ -18,16 +18,18 @@ taskwarrior-java-client
 What is it
 ----------
 
-A Java client to communicate with a [taskwarrior][] server (= [taskd](https://taskwarrior.org/docs/taskserver/why.html)).
+A Java client to communicate with a [taskwarrior][] server (=
```
Comment (ngortheone) [inline README.md:21]: broken long strings to 80 chars, easier to read in text editor - avoid wrapping

File: src/main/resources/taskwarrior.properties.template [hunk 7]
```diff
@@ -3,7 +3,6 @@ taskwarrior.server.host=
 taskwarrior.server.port=
 
 # Certificates and keys for SSL connection
-taskwarrior.ssl.folder=
```
Comment (ngortheone) [inline src/main/resources/taskwarrior.properties.template:6]: looks like this property is not used anywhere

**Comment by ngortheone** [comment]
@aaschmid 
Fixed all review comments.

> One remaining question: Where do you want to run this automatically? travis-ci, circle-ci?

I haven't thought much about this yet. We should probably try hook this up to existing Travis build (github now also has some CI, never tried it, but looks similar to travis). Not sure if Travis can build and run docker containers as part of the build process though. (Travis builds already run in docker, so that would require running docker in docker, and that requires privileged container)

Basic execution steps are:
- Build taskd container
- Run taskd container
- Run java tests
- Stop container

File: build.gradle [hunk 8]
```diff
@@ -43,13 +40,15 @@ jar {
     }
 }
 
+test {
+    useJUnitPlatform()
+}
+
 task javadocJar(type: Jar, dependsOn: javadoc) {
-    classifier = 'javadoc'
```
Comment (ngortheone) [inline build.gradle:47]: deprecated property

File: build.gradle [hunk 9]
```diff
@@ -43,13 +40,15 @@ jar {
...
     from javadoc.destinationDir
 }
 
 task sourcesJar(type: Jar) {
-    classifier = 'sources'
```
Comment (ngortheone) [inline build.gradle:52]: deprecated property

File: build.gradle [hunk 10]
```diff
@@ -1,25 +1,22 @@
 plugins {
     id 'de.aaschmid.cpd' version '1.0'
     id 'eclipse'
+    id 'idea'
```
Comment (aaschmid) [inline build.gradle]: This should not be required anymore, some for `eclipse`.

**Review by aaschmid — COMMENTED** [review]
nice, still wanna try it my own and maybe enhance travis build to be able to exeucte it automatically.

File: build.gradle (hunk 10 above)
Comment (ngortheone) [reply build.gradle]: ack. will remove

**Comment by ngortheone** [comment]
> nice, still wanna try it my own and maybe enhance travis build to be able to execute it automatically.

Sure. LMK if you want me to change/fix anything

**Comment by aaschmid** [comment]
Nice, the only missing thing would be, that the travis job runs green, see https://travis-ci.org/aaschmid/taskwarrior-java-client/builds/579532030?utm_source=github_status&utm_medium=notification. Would you mind to google the failure?
Maybe the only thing is to update `.travis.yml` to choose the correct `dist`, maybe `dist: trusty`...

**Comment by aaschmid** [comment]
@ngortheone Can you give me access to the source branch that I am able to adjust your changes and get it running? In top is stated that you deleted the repo?

image

Meanwhile I also changed the build circleci, see new branch https://github.com/aaschmid/taskwarrior-java-client/tree/bring.library.back.to.live

**Comment by ngortheone** [comment]
Hi @aaschmid ,

Sorry, I have abandoned my efforts in this field. I have started a total rewrite of your code to Kotlin, added more high level features. But then I life took an unexpected course...

Feel free to fork it or lift code from there.
https://github.com/ngortheone/paratrooper

**Comment by aaschmid** [comment]
Nevermind! I was a bit slow in my reaction :-) 
Finally I got your things merged and have adjusted to the needs. 

Your help and inspiration was very welcome, thanks again to you @ngortheone :-)

**Pull request merged**